src/prerender.ts

```diff
@@ -0,0 +1,339 @@
+				break;
+			}
+		}
+
+		if (!entryScript)
+			throw new Error("Unable to detect prerender entry script");
+
+		return path.join(viteConfig.root, entryScript);
+	};
+
+	return {
+		name: "headless-prerender",
+		apply: "build",
+		enforce: "post",
+		configResolved(config) {
+			viteConfig = config;
+		},
+		async options(opts) {
+			if (!opts.input) return;
+			if (!prerenderScript) {
+				prerenderScript = await getPrerenderScriptFromHTML(opts.input);
+			}
+
+			// prettier-ignore
+			opts.input =
+				typeof opts.input === "string"
+					? [opts.input, prerenderScript]
+					: Array.isArray(opts.input)
+						? [...opts.input, prerenderScript]
+						: { ...opts.input, prerenderEntry: prerenderScript };
+			opts.preserveEntrySignatures = "allow-extension";
+		},
+		// Injects a window check into Vite's preload helper, instantly resolving
+		// the module rather than attempting to add a <link> to the document.
+		transform(code, id) {
+			// Vite keeps changing up the ID, best we can do for cross-version
+			// compat is an `includes`
+			if (id.includes(preloadHelperId)) {
+				// Through v5.0.4
+				// https://github.com/vitejs/vite/blob/b93dfe3e08f56cafe2e549efd80285a12a3dc2f0/packages/vite/src/node/plugins/importAnalysisBuild.ts#L95-L98
+				const s = new MagicString(code);
+				s.replace(
+					`if (!__VITE_IS_MODERN__ || !deps || deps.length === 0) {`,
+					`if (!__VITE_IS_MODERN__ || !deps || deps.length === 0 || typeof window === 'undefined') {`,
+				);
+				// 5.0.5+
+				// https://github.com/vitejs/vite/blob/c902545476a4e7ba044c35b568e73683758178a3/packages/vite/src/node/plugins/importAnalysisBuild.ts#L93
+				s.replace(
+					`if (__VITE_IS_MODERN__ && deps && deps.length > 0) {`,
+					`if (__VITE_IS_MODERN__ && deps && deps.length > 0 && typeof window !== 'undefined') {`,
+				);
+				return {
+					code: s.toString(),
+					map: s.generateMap({ hires: true }),
+				};
+			}
+		},
+		async generateBundle(_opts, bundle) {
+			// @ts-ignore
+			globalThis.location = {};
+			// @ts-ignore
+			globalThis.self = globalThis;
+
+			// Local, fs-based fetch implementation for prerendering
+			const nodeFetch = globalThis.fetch;
+			// @ts-ignore
+			globalThis.fetch = async (url: string, opts: RequestInit | undefined) => {
+				if (/^\//.test(url)) {
+					const text = () =>
+						fs.readFile(
+							`${path.join(
+								viteConfig.root,
+								viteConfig.build.outDir,
+							)}/${url.replace(/^\//, "")}`,
+							"utf-8",
+						);
+					return { text, json: () => text().then(JSON.parse) };
+				}
+
+				return nodeFetch(url, opts);
+			};
+
+			// Grab the generated HTML file, which we'll use as a template:
+			const tpl = (bundle["index.html"] as OutputAsset).source as string;
+			let htmlDoc = htmlParse(tpl);
+
+			// Create a tmp dir to allow importing & consuming the built modules,
+			// before Rollup writes them to the disk
+			const tmpDir = await fs.mkdtemp(
+				path.join(os.tmpdir(), "headless-prerender-"),
+			);
```

Comment:
TODO: Bury this in `node_modules` I guess -- https://github.com/preactjs/preact/discussions/4228#discussioncomment-7842722